### PR TITLE
SpeechBrain pipeline for source separation inference

### PR DESCRIPTION
**Pull request by ycemsubakan**
Here's a commit for SpeechBrain source separation pipeline. 

The `__call__` function returns a numpy array of size `Ntimepoints x Nsources`. 

I am not sure if this would work for the huggingface's interface, but I wanted to ask via this PR.

### CODE REVIEW

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py [hunk 1]
```diff
@@ -22,9 +20,36 @@ def __call__(self, inputs: np.array) -> Tuple[np.array, int]:
...
+            resamp = torchaudio.transforms.Resample(
+                orig_freq=16000, new_freq=fs_model
+            )
+            mix = resamp(mix)
+
+        # separate
+        est_sources = self.model.separate_batch(mix)
+
+        # normalize for loudness
+        est_sources = est_sources / est_sources.max(dim=1, keepdim=True)[0]
```
Comment (Narsil) [inline api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py:47]: Shouldn't we use the first slice of the batch here already ?

```python
# Only 1 element in the batch is present
est_sources = est_sources[0]

#normalize for loudness
est_sources = est_sources / est_sources.max(dim=0, keepdim=True)

# return estimated sources as a T(time) x S(n.sources) matrix 
return est_sources.numpy(), fs_model
```

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py [hunk 2]
```diff
@@ -22,9 +20,36 @@ def __call__(self, inputs: np.array) -> Tuple[np.array, int]:
...
-        raise NotImplementedError(
-            "Please implement AudioSourceSeparationPipeline __call__ function"
-        )
+
+
+        fs_model = self.model.hparams.sample_rate
+
+        # we resample to the model's sampling frequency if needed
+        # we assume that the input is at 16kHz
+        mix = torch.from_numpy(inputs)
```
Comment (Narsil) [inline api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py:36]: The input is not batched, so should we make a fake 1-batch here maybe ?

Also question for the models, does stereo or multi channel audio make sense for some datasets/tasks/models ?

**Review by Narsil — COMMENTED** [review]
Hi @ycemsubakan .

This is exactly what should go there, thanks for this !

Unfortunately `audio-source-separation` is not as ironed as some other pipelines.
I'll help iron it out (by opening another PR and linking so you can track progress too).

Just a brief description:
- The pipeline should declare its sampling size, so **no** resampling should happen in the pipeline (ffmpeg handles it both ways if required and only once, that also take care of a very wide variety of audio formats that might be send towards the API)
- The output you have is currently a numpy array of TxS, it seems a bit more "natural" if it was a list of T arrays (there is an ongoing discussion internally about the tasks and the various input output) but we can use this for now (the `api-inference-community` package adds a layer on top that could handle that potential future change for you.
- Currently the input is `[T]` array as sampling rate `AudioSourceSeparationPipeline.sampling_rate`, will it ever make sense to receive multiple audio streams (I could imagine stereo sound being helpful to do source separation) ?

Pinging @osanseviero into the discussion.

**Comment by ycemsubakan** [comment]
> Hi @ycemsubakan .
> 
> This is exactly what should go there, thanks for this !
> 
> Unfortunately `audio-source-separation` is not as ironed as some other pipelines.
> I'll help iron it out (by opening another PR and linking so you can track progress too).
> 
> Just a brief description:
> 
>     * The pipeline should declare its sampling size, so **no** resampling should happen in the pipeline (ffmpeg handles it both ways if required and only once, that also take care of a very wide variety of audio formats that might be send towards the API)
> 
>     * The output you have is currently a numpy array of TxS, it seems a bit more "natural" if it was a list of T arrays (there is an ongoing discussion internally about the tasks and the various input output) but we can use this for now (the `api-inference-community` package adds a layer on top that could handle that potential future change for you.
> 
>     * Currently the input is `[T]` array as sampling rate `AudioSourceSeparationPipeline.sampling_rate`, will it ever make sense to receive multiple audio streams (I could imagine stereo sound being helpful to do source separation) ?
> 
> 
> Pinging @osanseviero into the discussion.

Hi @Narsil , thanks for the quick reply. Here are my answers:
 * The thing is, some models are trained with 8k and some others with 16k. Would it be possible to set the pipeline's sampling frequency depending on the model? 
 * Then we can keep the `TxS` matrix as output for now I guess. 
 * At the moment we don't have stereo models, but yes, we might have stereo inputs. For that I guess we can use a matrix of size `Tx2`, which doesn't change the pipeline I think?

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py (hunk 2 above)
Comment (ycemsubakan) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py:36]: Ah ok, yes we should do `.unsqueeze(0)` to introduce a fake batch dimension.. 
Stereo or multichannel might make sense, but the input can be taken as a tensor in that case also, right?

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py (hunk 1 above)
Comment (ycemsubakan) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py:47]: Sure, these changes work for me..

Comment (ycemsubakan) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_source_separation.py]: I commited the suggested changes.

**Comment by Narsil** [comment]
> > Hi @ycemsubakan .
> > This is exactly what should go there, thanks for this !
> > Unfortunately `audio-source-separation` is not as ironed as some other pipelines.
> > I'll help iron it out (by opening another PR and linking so you can track progress too).
> > Just a brief description:
> > ```
> > * The pipeline should declare its sampling size, so **no** resampling should happen in the pipeline (ffmpeg handles it both ways if required and only once, that also take care of a very wide variety of audio formats that might be send towards the API)

> 
> Hi @Narsil , thanks for the quick reply. Here are my answers:
> 
>     * The thing is, some models are trained with 8k and some others with 16k. Would it be possible to set the pipeline's sampling frequency depending on the model?

Yes, it is currently done with AutomaticSpeechRecognition for instance, it's a required attribute on the pipeline class.
That way the validation can read it and sample the input audio accordingly.

> 
>     * Then we can keep the `TxS` matrix as output for now I guess.
> 
>     * At the moment we don't have stereo models, but yes, we might have stereo inputs. For that I guess we can use a matrix of size `Tx2`, which doesn't change the pipeline I think?

Well it does change the input shape of the audio array received. So it's not a huge deal, but still important because the less breaking change we introduce the better.

**Comment by ycemsubakan** [comment]
Hi @Narsil I was just wondering if you need anything from me for this PR?

**Comment by Narsil** [comment]
@ycemsubakan I opened another PR with you referenced as author (as the main contribution is from you, I merely changed some sugar on top) as we discussed changing a bit to avoid having too many pipeline names:

https://github.com/huggingface/huggingface_hub/pull/94

Feel free to comment there. We were thinking of merging this soon (hence enabling all new speechbrain models on the API) @osanseviero

**Comment by ycemsubakan** [comment]
Thanks @Narsil !

**Comment by Narsil** [comment]
I will close this as this has been superseeded. Everything will be online soon.

**Pull request closed without merging**